gh-132108: Add Buffer Protocol support to int.from_bytes to improve performance

cmaloney opened the pull request:
Speed up conversion from `bytes-like` objects like `bytearray` while keeping conversion from `bytes` stable.

On a `--with-lto --enable-optimizaitons` build on my 64 bit Linux box:

new:
```
from_bytes_flags: Mean +- std dev: 28.6 ns +- 0.5 ns
bench_convert[bytes]: Mean +- std dev: 50.4 ns +- 1.4 ns
bench_convert[bytearray]: Mean +- std dev: 51.3 ns +- 0.7 ns
```

old:
```
from_bytes_flags: Mean +- std dev: 28.1 ns +- 1.1 ns
bench_convert[bytes]: Mean +- std dev: 50.3 ns +- 4.3 ns
bench_convert[bytearray]: Mean +- std dev: 64.7 ns +- 0.9 ns
```

Benchmark code:
```python
import pyperf
import time

def from_bytes_flags(loops):
    range_it = range(loops)

    t0 = time.perf_counter()
    for _ in range_it:
        int.from_bytes(b'\x00\x10', byteorder='big')
        int.from_bytes(b'\x00\x10', byteorder='little')
        int.from_bytes(b'\xfc\x00', byteorder='big', signed=True)
        int.from_bytes(b'\xfc\x00', byteorder='big', signed=False)
        int.from_bytes([255, 0, 0], byteorder='big')
    return time.perf_counter() - t0

sample_bytes = [
    b'',
    b'\x00',
    b'\x01',
    b'\x7f',
    b'\x80',
    b'\xff',
    b'\x01\x00',
    b'\x7f\xff',
    b'\x80\x00',
    b'\xff\xff',
    b'\x01\x00\x00',
]

sample_bytearray = [bytearray(v) for v in sample_bytes]

def bench_convert(loops, values):
    range_it = range(loops)

    t0 = time.perf_counter()
    for _ in range_it:
        for val in values:
            int.from_bytes(val)
    return time.perf_counter() - t0

runner = pyperf.Runner()

runner.bench_time_func('from_bytes_flags', from_bytes_flags, inner_loops=10)
runner.bench_time_func('bench_convert[bytes]', bench_convert, sample_bytes, inner_loops=10)
runner.bench_time_func('bench_convert[bytearray]', bench_convert, sample_bytearray, inner_loops=10)
```


<!-- gh-issue-number: gh-132108 -->
* Issue: gh-132108
<!-- /gh-issue-number -->

cmaloney committed:
gh-132108: Add Buffer Protocol support to int.from_bytes

Speed up conversion from `bytes-like` objects like `bytearray` while
keeping conversion from `bytes` stable.

On a `--with-lto --enable-optimizaitons` build on my 64 bit Linux box:

new:
from_bytes_flags: Mean +- std dev: 28.6 ns +- 0.5 ns
bench_convert[bytes]: Mean +- std dev: 50.4 ns +- 1.4 ns
bench_convert[bytearray]: Mean +- std dev: 51.3 ns +- 0.7 ns

old:
from_bytes_flags: Mean +- std dev: 28.1 ns +- 1.1 ns
bench_convert[bytes]: Mean +- std dev: 50.3 ns +- 4.3 ns
bench_convert[bytearray]: Mean +- std dev: 64.7 ns +- 0.9 ns

Benchmark code:
```python
import pyperf
import time

def from_bytes_flags(loops):
    range_it = range(loops)

    t0 = time.perf_counter()
    for _ in range_it:
        int.from_bytes(b'\x00\x10', byteorder='big')
        int.from_bytes(b'\x00\x10', byteorder='little')
        int.from_bytes(b'\xfc\x00', byteorder='big', signed=True)
        int.from_bytes(b'\xfc\x00', byteorder='big', signed=False)
        int.from_bytes([255, 0, 0], byteorder='big')
    return time.perf_counter() - t0

sample_bytes = [
    b'',
    b'\x00',
    b'\x01',
    b'\x7f',
    b'\x80',
    b'\xff',
    b'\x01\x00',
    b'\x7f\xff',
    b'\x80\x00',
    b'\xff\xff',
    b'\x01\x00\x00',
]

sample_bytearray = [bytearray(v) for v in sample_bytes]

def bench_convert(loops, values):
    range_it = range(loops)

    t0 = time.perf_counter()
    for _ in range_it:
        for val in values:
            int.from_bytes(val)
    return time.perf_counter() - t0

runner = pyperf.Runner()

runner.bench_time_func('from_bytes_flags', from_bytes_flags, inner_loops=10)
runner.bench_time_func('bench_convert[bytes]', bench_convert, sample_bytes, inner_loops=10)
runner.bench_time_func('bench_convert[bytearray]', bench_convert, sample_bytearray, inner_loops=10)
```

picnixz reviewed (dismissed):
Can we have benchmarks for very large bytes? maybe you can also say how much we're gaining in the NEWS entry that way.

sobolevn commented on Misc/NEWS.d/next/Core_and_Builtins/2025-04-04-20-38-29.gh-issue-132108.UwZIQy.rst:
```diff
@@ -0,0 +1,2 @@
+Speed up :meth:`int.from_bytes` when passed a bytes-like object such as a
+:class:`bytearray`.
```
```suggestion
Speed up :meth:`int.from_bytes` when passed a bytes-like object such as
:class:`bytes` and :class:`bytearray`.
```

serhiy-storchaka replied:
I do not think that it affects `bytes`.

sobolevn reviewed (commented):
This is a breaking change. Example.

Before:

```python
>>> class X(bytes):
...     def __bytes__(self):
...         return b'X'
...         
>>> int.from_bytes(X(b'a'))
88
```

After:

```python
>>> class X(bytes):
...     def __bytes__(self):
...         return b'X'
...         
>>> int.from_bytes(X(b'a'))
97
```

skirpichev commented on Misc/NEWS.d/next/Core_and_Builtins/2025-04-04-20-38-29.gh-issue-132108.UwZIQy.rst:
```diff
@@ -0,0 +1,2 @@
+Speed up :meth:`int.from_bytes` when passed a bytes-like object such as a
+:class:`bytearray`.
```
```suggestion
Speed up :meth:`int.from_bytes` when passed object supports :ref:`buffer
protocol <bufferobjects>`, like :class:`bytearray`.
```

serhiy-storchaka reviewed (approved):
LGTM.

cmaloney committed:
Apply suggestions from code review

Co-authored-by: Sergey B Kirpichev <skirpichev@gmail.com>

cmaloney committed:
Check bytes exact first to speed up that case

from_bytes_flags: Mean +- std dev: [main] 28.3 ns +- 1.3 ns -> [exactbytes] 27.3 ns +- 0.3 ns: 1.04x faster
bench_convert[bytearray]: Mean +- std dev: [main] 65.8 ns +- 3.3 ns -> [exactbytes] 53.1 ns +- 5.1 ns: 1.24x faster
bench_convert_big[bytes]: Mean +- std dev: [main] 51.8 ns +- 0.6 ns -> [exactbytes] 50.3 ns +- 0.5 ns: 1.03x faster
bench_convert_big[bytearray]: Mean +- std dev: [main] 65.8 ns +- 3.0 ns -> [exactbytes] 53.5 ns +- 5.3 ns: 1.23x faster

Benchmark hidden because not significant (1): bench_convert[bytes]

cmaloney committed:
Tweak comments, fix PEP-7 formatting

gpshead commented on Objects/longobject.c:
```diff
@@ -6452,14 +6453,32 @@ int_from_bytes_impl(PyTypeObject *type, PyObject *bytes_obj,
+    }
+    /* Use buffer protocol to avoid copies. */
+    else if (PyObject_CheckBuffer(bytes_obj)) {
+        if (PyObject_GetBuffer(bytes_obj, &view, PyBUF_SIMPLE) != 0) {
```
declare Py_buffer view within this block since it is the only place it is used.

cmaloney replied:
Fixed :)

gpshead reviewed (approved):
I think we could go ahead with this PR even while waiting on the larger discussion as it does make sense.  Regardless, if you want to do a simpler version that just adds the PyBytes_CheckExact optimization for starters, also feel free.  that is a pure optimization of existing behavior.

vstinner reviewed (approved):
LGTM. The latest PR is now backward compatible.

```py
>>> class X(bytes):
...     def __bytes__(self):
...         return b'X'
...         
>>> int.from_bytes(X(b'a'))
97
```

cmaloney committed:
address review comment: Move Py_buffer view into block

vstinner committed:
Merge branch 'main' into int_from_byteslike
